### PR TITLE
Use decorated serializer to resolve url of uploaded files

### PR DESCRIPTION
**Pull request by lukasluecke**
This matches the recommended approach of using our extension points, and also works for GraphQL

### CODE REVIEW

File: core/file-upload.md [hunk 1]
```diff
@@ -164,64 +164,65 @@ final class CreateMediaObjectAction
...
+    public function supportsDenormalization($data, $type, $format = null)
+    {
+        return $this->decorated->supportsDenormalization($data, $type, $format);
+    }
 
-        foreach ($mediaObjects as $mediaObject) {
-            if (!$mediaObject instanceof MediaObject) {
-                continue;
-            }
+    public function denormalize($data, $class, $format = null, array $context = [])
```
Comment (alanpoulain) [inline core/file-upload.md]: I don't think we need to denormalize a `MediaObject`. You can only use the `NormalizerInterface` IMO.

File: core/file-upload.md [hunk 2]
```diff
@@ -164,64 +164,65 @@ final class CreateMediaObjectAction
...
+    {
+        return $this->decorated->denormalize($data, $class, $format, $context);
+    }
 
-            $mediaObject->contentUrl = $this->storage->resolveUri($mediaObject, 'file');
+    public function setSerializer(SerializerInterface $serializer)
+    {
+        if($this->decorated instanceof SerializerAwareInterface) {
```
Comment (alanpoulain) [inline core/file-upload.md]:
```suggestion
        if ($this->decorated instanceof SerializerAwareInterface) {
```
I know this CS is in the other part as well.

File: core/file-upload.md (hunk 1 above)
Comment (lukasluecke) [reply core/file-upload.md]: You're probably right - I just took the example from the Serialization docs and modified it accordingly. Will update this 👍

File: core/file-upload.md [hunk 3]
```diff
@@ -164,64 +164,65 @@ final class CreateMediaObjectAction
...
-use Symfony\Component\HttpKernel\Event\GetResponseForControllerResultEvent;
-use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerInterface;
 use Vich\UploaderBundle\Storage\StorageInterface;
 
-final class ResolveMediaObjectContentUrlSubscriber implements EventSubscriberInterface
+final class MediaObjectContentUrlNormalizer implements NormalizerInterface, DenormalizerInterface, SerializerAwareInterface
```
Comment (teohhanhui) [inline core/file-upload.md]: Just name it `MediaObjectNormalizer` :smile:

File: core/file-upload.md (hunk 1 above)
Comment (teohhanhui) [reply core/file-upload.md]: The problem is, if we're declaring this as a decorating service in Symfony DI, then it should implement the same interfaces. I'd suggest that we do **NOT** declare this as a decorating service (i.e. no `decorates`).

This way we can also remove `SerializerAwareInterface`.

Comment (lukasluecke) [reply core/file-upload.md]: @teohhanhui Sounds good - does that change the behavior in any way? Is there any additional configuration required if you do not decorate the existing service?

File: core/file-upload.md [hunk 4]
```diff
@@ -164,64 +164,65 @@ final class CreateMediaObjectAction
 Returning the plain file path on the filesystem where the file is stored is not useful for the client, which needs a
 URL to work with.
 
-An [event subscriber](events.md#custom-event-listeners) could be used to set the `contentUrl` property:
+We will be [decorating a serializer](serialization.md#decorating-a-serializer-and-adding-extra-data) to resolve the `contentUrl` property:
```
[the remark this answers is not in the document]
Comment (lukasluecke) [reply core/file-upload.md]: @teohhanhui What do you mean? I took both the title and the link from the existing documentation.

(If we move away from the decoration, then this section has to be updated anyway.)

File: core/file-upload.md (hunk 3 above)
Comment (lukasluecke) [reply core/file-upload.md]: I tried to match the previous name somewhat, but shortening it is fine with me.

File: core/file-upload.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dunglas) [reply core/file-upload.md]: It could be renamed, but in another PR as here it's just a link.

[the remark this answers is not in the document]
Comment (dunglas) [reply core/file-upload.md]: Let's rename then!

File: core/file-upload.md (hunk 3 above)
Comment (dunglas) [reply core/file-upload.md]: Let's rename then!

**Comment by alanpoulain** [comment]
Superseded by https://github.com/api-platform/docs/pull/1339. Thank you @lukasluecke.

**Pull request closed without merging**